### PR TITLE
[5.1] WebAsset fix for calling the onAttachCallback, and cache for tree sorting to improve performance

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes

When asset implements `WebAssetAttachBehaviorInterface`  and added while `onBeforeCompilehead` event, then `onAttachCallback` was ignored.
The PR is fixing it. 

### Testing Instructions

Add following code to `onBeforeCompileHead` of any plugin (example in to debug plugin):
```php
$wa = \Joomla\CMS\Factory::getApplication()->getDocument()->getWebAssetManager();
$wa->registerAndUseScript(new class ('foobar') extends \Joomla\CMS\WebAsset\WebAssetItem implements \Joomla\CMS\WebAsset\WebAssetAttachBehaviorInterface
{
    public function onAttachCallback(\Joomla\CMS\Document\Document $doc){
        dd('FooBar');
    }
});
```

### Actual result BEFORE applying this Pull Request
Nothing happen

### Expected result AFTER applying this Pull Request
you get an empty page with "FooBar" text

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed
- [x] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

Reference:

https://manual.joomla.org/docs/next/general-concepts/web-asset-manager#advanced-custom-webassetitem-class

### CODE REVIEW

**Comment by martin-zw** [comment]
I have tested this item :white_check_mark: successfully on 2d1475d218d4c8223f040e37f6ea4fd8d28101dd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42661">issues.joomla.org/tracker/joomla-cms/42661</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 2d1475d218d4c8223f040e37f6ea4fd8d28101dd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42661">issues.joomla.org/tracker/joomla-cms/42661</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42661">issues.joomla.org/tracker/joomla-cms/42661</a>.</sub>

File: libraries/src/Document/Renderer/Html/MetasRenderer.php [hunk 1]
```diff
@@ -51,33 +51,41 @@ public function render($head, $params = [], $content = null)
...
         $wa  = $this->_doc->getWebAssetManager();
 
-        // Check for AttachBehavior and web components
-        foreach ($wa->getAssets('script', true) as $asset) {
-            if ($asset instanceof WebAssetAttachBehaviorInterface) {
-                $asset->onAttachCallback($this->_doc);
-            }
-        }
+        // Add a dummy asset for script options, this will prevent WebAssetManager from extra re-calculation later on.
+        $soAsset = $wa->addInline('script', '', ['name' => 'joomla.script.options'], [], ['core'])
```
Comment (LadySolveig) [inline libraries/src/Document/Renderer/Html/MetasRenderer.php]:
```suggestion
        $scriptOptionsAsset = $wa->addInline('script', '', ['name' => 'joomla.script.options'], [], ['core'])
```

File: libraries/src/Document/Renderer/Html/MetasRenderer.php [hunk 2]
```diff
@@ -51,33 +51,41 @@ public function render($head, $params = [], $content = null)
...
 
+        // Re-Check for AttachBehavior for newly added assets
+        WebAssetManager::callOnAttachCallback($wa->getAssets('script', true), $this->_doc, $onAttachCallCache);
+
         // Add Script Options as inline asset
         $scriptOptions = $this->_doc->getScriptOptions();
 
         if ($scriptOptions) {
+            // Overriding ScriptOptions asset is not allowed
+            if ($soAsset !== $wa->getAsset('script', 'joomla.script.options')) {
```
Comment (LadySolveig) [inline libraries/src/Document/Renderer/Html/MetasRenderer.php]:
```suggestion
            if ($scriptOptionsAsset !== $wa->getAsset('script', 'joomla.script.options')) {
```

File: libraries/src/Document/Renderer/Html/MetasRenderer.php [hunk 3]
```diff
@@ -51,33 +51,41 @@ public function render($head, $params = [], $content = null)
...
-                ['name' => 'joomla.script.options', 'position' => 'before'],
-                ['type' => 'application/json', 'class' => 'joomla-script-options new'],
-                ['core']
-            );
+
+            // Set content and update attributes of dummy asset to correct ones
+            $soAsset->setOption('content', $jsonOptions ?: '{}');
+            $soAsset->setOption('position', 'before');
+            $soAsset->setAttribute('type', 'application/json');
+            $soAsset->setAttribute('class', 'joomla-script-options new');
```
Comment (LadySolveig) [inline libraries/src/Document/Renderer/Html/MetasRenderer.php]:
```suggestion
            $scriptOptionsAsset->setAttribute('class', 'joomla-script-options new');
```

File: libraries/src/Document/Renderer/Html/MetasRenderer.php [hunk 4]
```diff
@@ -51,33 +51,41 @@ public function render($head, $params = [], $content = null)
...
-                $jsonOptions,
-                ['name' => 'joomla.script.options', 'position' => 'before'],
-                ['type' => 'application/json', 'class' => 'joomla-script-options new'],
-                ['core']
-            );
+
+            // Set content and update attributes of dummy asset to correct ones
+            $soAsset->setOption('content', $jsonOptions ?: '{}');
+            $soAsset->setOption('position', 'before');
+            $soAsset->setAttribute('type', 'application/json');
```
Comment (LadySolveig) [inline libraries/src/Document/Renderer/Html/MetasRenderer.php]:
```suggestion
            $scriptOptionsAsset->setAttribute('type', 'application/json');
```

File: libraries/src/Document/Renderer/Html/MetasRenderer.php [hunk 5]
```diff
@@ -51,33 +51,41 @@ public function render($head, $params = [], $content = null)
...
-            $wa->addInlineScript(
-                $jsonOptions,
-                ['name' => 'joomla.script.options', 'position' => 'before'],
-                ['type' => 'application/json', 'class' => 'joomla-script-options new'],
-                ['core']
-            );
+
+            // Set content and update attributes of dummy asset to correct ones
+            $soAsset->setOption('content', $jsonOptions ?: '{}');
+            $soAsset->setOption('position', 'before');
```
Comment (LadySolveig) [inline libraries/src/Document/Renderer/Html/MetasRenderer.php]:
```suggestion
            $scriptOptionsAsset->setOption('position', 'before');
```

File: libraries/src/Document/Renderer/Html/MetasRenderer.php [hunk 6]
```diff
@@ -51,33 +51,41 @@ public function render($head, $params = [], $content = null)
...
-
-            $wa->addInlineScript(
-                $jsonOptions,
-                ['name' => 'joomla.script.options', 'position' => 'before'],
-                ['type' => 'application/json', 'class' => 'joomla-script-options new'],
-                ['core']
-            );
+
+            // Set content and update attributes of dummy asset to correct ones
+            $soAsset->setOption('content', $jsonOptions ?: '{}');
```
Comment (LadySolveig) [inline libraries/src/Document/Renderer/Html/MetasRenderer.php]:
```suggestion
            $scriptOptionsAsset->setOption('content', $jsonOptions ?: '{}');
```

File: libraries/src/WebAsset/WebAssetManager.php [hunk 7]
```diff
@@ -991,4 +1016,33 @@ protected function getDependenciesForAsset(
...
+     * @since __DEPLOY_VERSION__
+     */
+    public static function callOnAttachCallback(array $assets, Document $document, array $cache = []): array
+    {
+        foreach ($assets as $asset) {
+            if (!$asset instanceof WebAssetAttachBehaviorInterface) {
+                continue;
+            }
+
+            $oid = spl_object_id($asset);
```
Comment (LadySolveig) [inline libraries/src/WebAsset/WebAssetManager.php:1038]: Would be glad to have here more self explantory and pronounceable variable names.

File: libraries/src/WebAsset/WebAssetManager.php [hunk 8]
```diff
@@ -991,4 +1016,33 @@ protected function getDependenciesForAsset(
...
+
+            if (empty($cache[$oid])) {
```
Comment (LadySolveig) [inline libraries/src/WebAsset/WebAssetManager.php:1040]: see above line 1038

File: libraries/src/WebAsset/WebAssetManager.php [hunk 9]
```diff
@@ -991,4 +1016,33 @@ protected function getDependenciesForAsset(
...
+                $asset->onAttachCallback($document);
+                $cache[$oid] = true;
```
Comment (LadySolveig) [inline libraries/src/WebAsset/WebAssetManager.php:1042]: see above line 1038

File: libraries/src/WebAsset/WebAssetManager.php (hunk 7 above)
Comment (Fedik) [reply libraries/src/WebAsset/WebAssetManager.php:1038]: it is ID of Object, no reason to go fancy here, it is well known shortening :wink:

File: libraries/src/Document/Renderer/Html/MetasRenderer.php (hunk 1 above)
Comment (Fedik) [reply libraries/src/Document/Renderer/Html/MetasRenderer.php]: I would prefer to keep shorter version, but okay

**Comment by LadySolveig** [comment]
Thank you @Fedik and for testing @martin-zw @viocassel

**Pull request merged**